File: codes/rust/42.接雨水.rs
/*
 * @lc app=leetcode.cn id=42 lang=rust
 *
 * [42] 接雨水
 */

// @lc code=start
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let mut res = 0;
        let mut left_max_memo = vec![0; height.len()];
        let mut right_max_memo = vec![0; height.len()];
        for i in 1..height.len() - 1 {
            left_max_memo[i] = left_max_memo[i - 1].max(height[i - 1]);
        }
        for i in (1..height.len() - 1).rev() {
            right_max_memo[i] = right_max_memo[i + 1].max(height[i + 1]);
        }
        for i in 1..height.len() - 1 {
            let min = left_max_memo[i].min(right_max_memo[i]);
            if min > height[i] {
                res += min - height[i];
            }
        }
        res
    }
}
// @lc code=end

File: codes/golang/42.接雨水.go


func trap(height []int) int {
	if len(height) == 0 {
		return 0
	}
	n := len(height)
	res := 0
	// 数组充当备忘录
	l_max := make([]int, n)
	r_max := make([]int, n)
	// 初始化 base case
	l_max[0] = height[0]
	r_max[n-1] = height[n-1]
	// 从左向右计算 l_max
	for i := 1; i < n; i++ {
		l_max[i] = max(height[i], l_max[i-1])
	}
	// 从右向左计算 r_max
	for i := n - 2; i >= 0; i-- {
		r_max[i] = max(height[i], r_max[i+1])
	}
	// 计算答案
	for i := 1; i < n-1; i++ {
		res += min(l_max[i], r_max[i]) - height[i]
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}